clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,69 +370,116 @@ public void poll(long timeout) throws IOException {
+                            SocketChannel socketChannel = (SocketChannel) key.channel();
+                            log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
+                                    socketChannel.socket().getReceiveBufferSize(),
+                                    socketChannel.socket().getSendBufferSize(),
+                                    socketChannel.socket().getSoTimeout(),
+                                    channel.id());
+                        } else
+                            continue;
+                    }
 
-                /* complete any connections that have finished their handshake (either normally or immediately) */
-                if (isImmediatelyConnected || key.isConnectable()) {
-                    if (channel.finishConnect()) {
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
-                        SocketChannel socketChannel = (SocketChannel) key.channel();
-                        log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
-                                socketChannel.socket().getReceiveBufferSize(),
-                                socketChannel.socket().getSendBufferSize(),
-                                socketChannel.socket().getSoTimeout(),
-                                channel.id());
-                    } else
-                        continue;
-                }
+                    /* if channel is not ready finish prepare */
+                    if (channel.isConnected() && !channel.ready()) {
+                        madeProgress = true;
+                        channel.prepare();
+                    }
 
-                /* if channel is not ready finish prepare */
-                if (channel.isConnected() && !channel.ready())
-                    channel.prepare();
+                    /* if channel is ready read from any connections that have readable data */
+                    if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
+                        NetworkReceive networkReceive;
+                        while ((networkReceive = channel.read()) != null) {
+                            madeProgress = true;
+                            addToStagedReceives(channel, networkReceive);
+                        }
+                        if (channel.hasBytesBuffered()) {
+                            //this channels has bytes enqueued in intermediary buffers that we could not read
+                            //(possibly because no memory). it may be the case that the underlying socket will
+                            //not come up in the next poll() and so we need to remember this channel for the
+                            //next poll call otherwise data may be stuff in said buffers forever.
+                            keysWithBytesBuffered.add(key);
+                        }
```

Comment:
Since`keysWithBytesBuffered` was cleared earlier, it needs to be populated regardless of the status of staged receives. I think `"if(..) { keysWithBytesBuffered.add(..); }"` should be done outside the outer if that checks staged receives.